### DISCUSSION TITLE
How can I make a table column shrink to fit the longest width?

### DISCUSSION THREAD

**Discussion by Hypersapien** [QuestPDF/QuestPDF#517] [General]
If I have a table with multiple rows, how can I make a column (say, the first column) shrink to fit the longest text in any of the cells without wrapping the text or specifying a constant width for the column?

**Comment by flew2bits** [comment]
This is not possible. Doing so would require measuring the width of the text in all rows of the table before rendering in order to calculate the widths of all of the columns.

The column widths are pre-computed based on the available width of the table's container. Then based on those now-fixed column widths, it determines the height needed per row.

It might be possible to modify the code to allow one or two "auto" columns, but you'd risk running into a layout exception if you had unexpectedly long text. This would force any other columns to be too small to hold content.

**Reply by Hypersapien** [reply]
Ok. Is there a way to find out how long a particular string is going to be in the given font and font size so we can programmatically set the size of the column?

**Reply by flew2bits** [reply]
The ability to measure content is not available in the exposed API except through the DynamicComponent, but I don't think that would be helpful in this situation.

**Reply by Hypersapien** [reply]
How would it not be helpful? If you have a piece of text that's generated by a variable so you don't know exactly how big it's going to be, you have have your code look at how long it will be in the final PDF and adjust the layout accordingly.

**Reply by flew2bits** [reply]
After thinking about it for a bit, it might be possible to do this with the DynamicComponent after all... You'd have to do it in a manner similar to the table example here https://www.questpdf.com/concepts/dynamic-components.html#optimized-example. I don't know the exact details (I'd have to experiment), but you'd first have to measure the size of each of the rows of data in the automatically sized column(s). When creating the rows in the subsequent runs, you'd just use that in the column definitions with a ConstantItem.

You can also use the non-optimized version, but you'd probably more benefit from the optimized version.

**Reply by flew2bits** [reply]
Here's a modified version of the optimized DynamicComponent that pre-calculates the width of the column. I added the DebugArea call to show that the column is the width of the longest text.

https://gist.github.com/flew2bits/8ca9647e838790a69f88873d8c0a9389

**Reply by flew2bits** [reply]
The `CreateElement` method uses `Size.Max` when measuring the content, so you can count on it rendering the text on a single line. This means that if the calculated `AutoWidth` is greater than the `context.AvailableSize.Width`, you'll get a layout error because the column won't fit on the page.

**Reply by Hypersapien** [reply]
I can't figure out where to get the DynamicContext from to pass it in.

**Comment by flew2bits** [comment]
Here's a branch that I created that hacks an auto-column option in to the column definition. During the draw stage, QuestPDF updates the column widths, but the Cells enumerable is empty, so it has to skip over the text measurement during that phase.

I don't know that this is the best way to handle this, but it does work in my testing.

https://github.com/flew2bits/QuestPDF/tree/table-auto-column

**Reply by flew2bits** [reply]
Here's the diff to see the changes I made: https://github.com/QuestPDF/QuestPDF/compare/main...flew2bits:QuestPDF:table-auto-column

**Reply by flew2bits** [reply]
Oh, and it won't handle spanned rows or columns. That would just overly complicate things.

**Reply by flew2bits** [reply]
Keep in mind that with this particular hack, if a single cell in the auto column equals or exceeds the available width, you will get a layout error. It assumes that you want the text to fit on a single line. You could probably add an option to allow multiple lines, but that could get tricky, fast.

**Reply by tn-5** [reply]
I had a similar issue and solved it by measuring all the cells & setting the column width to the max measured cell width per column. It is for smallish data sets (max couple of 100 rows with 20 to 30 columns) but the performance is more than ok for my purpose. 

I used a modified source version and did this calculation in the `UpdateColumnsWidth` function of the `Table`.  The `TableColumnDefinition` is also a bit modified - MinWidth + MaxWidth supported with autofit to fit contents. It is a work in progress so there will obviously be cases where it doesn't work properly but in principle it is possible.

**Comment by Lure5134** [comment]
@MarcinZiabek In my project, I would also need columns that automatically adjust the width. My suggestion would be to add 2 new properties to the columns. AllowShrink and AllowGrow. If the table is created, you could first calculate the widths as they are currently calculated. Then you would go through all the columns that have AllowShrink set to true and shrink the columns (if possible). Then you would widen all columns (if any) that have AllowGrow set to true until the width of the columns matches the width of the table. Could I create a PR in which I implement this for the ConstantColumn as well as for the RelativeColumn? And would you then also look at the PR?

**Reply by MarcinZiabek** [reply]
Thank you for sharing your idea; I greatly appreciate it 😄 

Indeed, your proposition may work quite well. I am mostly worried about: 
- performance - to determine the size of grow/shrink columns, every element needs to be measured with infinite space. It will at least double the execution time,
- predictiveness - the grow/shrink column may request more horizontal space than available, leading to layout exceptions. Using the `MaxWidth` may be necessary in may situations to prevent such cases,
- corner cases (the most problematic) - adjusting columns may be difficult when you have content that spans multiple columns/rows. Some cells can even partially overlap each other,
- paging support - many elements change size requirements depending on available space (e.g. `AspectRatio`), and therefore may produce unstable layouts.

I am not against giving it a try. I am just aware that it may greatly complicate this layout structure, which frankly is already quite complex. If you have any ideas, please feel free to share.

**Reply by Lure5134** [reply]
I have made here a draft pull request: https://github.com/QuestPDF/QuestPDF/pull/867
This currently only works for cells that don't use row or column spans.
The case that the columns require more space than is available should not occur, as the columns are only enlarged if a column is also reduced in size and the maximum amount of free space that is freed up by the reduction in size can be used.
